compactor/src/compact.rs

```diff
@@ -167,47 +240,399 @@ impl Compactor {
         }
         tombstones
     }
-}
 
-#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord)]
-struct TablePartition {
-    sequencer_id: SequencerId,
-    table_id: TableId,
-    partition_id: PartitionId,
-}
+    // Compact given files. Assume the given files are overlaped in time.
+    // If the assumption does not meet, we will spend time not to compact anything but put data together
+    async fn compact(
+        &self,
+        overlapped_files: Vec<ParquetFileWithTombstone>,
+    ) -> Result<Option<CompactedData>> {
+        // Nothing to compact
+        if overlapped_files.is_empty() {
+            return Ok(None);
+        }
 
-impl TablePartition {
-    fn new(sequencer_id: SequencerId, table_id: TableId, partition_id: PartitionId) -> Self {
-        Self {
-            sequencer_id,
-            table_id,
-            partition_id,
+        // One file without tombstone, no need to compact
+        if overlapped_files.len() == 1 && overlapped_files[0].tombstones.is_empty() {
+            return Ok(None);
         }
-    }
-}
 
-struct ParquetFileWithTombstone {
-    data: Arc<ParquetFile>,
-    tombstones: Vec<Tombstone>,
-}
+        // Keep the fist IoxMetadata to reuse same IDs and names
+        let iox_metadata = overlapped_files[0].iox_metadata();
+
+        // Verify if the given files belong to the same partition
+        // Note: we can ignore this verification if we assume this is a must-have condition
+        if overlapped_files.len() > 1 {
+            for file in &overlapped_files {
+                if file.data.sequencer_id != overlapped_files[0].data.sequencer_id
+                    || file.data.table_id != overlapped_files[0].data.table_id
+                    || file.data.partition_id != overlapped_files[0].data.partition_id
+                {
+                    return Err(Error::NotSamePartition {
+                        sequencer_id_1: overlapped_files[0].data.sequencer_id,
+                        table_id_1: overlapped_files[0].data.table_id,
+                        partition_id_1: overlapped_files[0].data.partition_id,
+                        sequencer_id_2: file.data.sequencer_id,
+                        table_id_2: file.data.table_id,
+                        partition_id_2: file.data.partition_id,
+                    });
+                }
+            }
+        }
```

Comment:
The current code checks the first element against itself and can be simplified

```suggestion
        if let Some((head, tail)) = overlapped_files.split_first() {
            for file in tail {
                let is_same = file.data.sequencer_id == head.data.sequencer_id
                    && file.data.table_id == head.data.table_id
                    && file.data.partition_id == head.data.partition_id;

                ensure!(is_same, NotSamePartitionSnafu {
                    sequencer_id_1: head.data.sequencer_id,
                    table_id_1: head.data.table_id,
                    partition_id_1: head.data.partition_id,
                    sequencer_id_2: file.data.sequencer_id,
                    table_id_2: file.data.table_id,
                    partition_id_2: file.data.partition_id,
                })
            }
        }
```